File: mandi-direct/frontend/src/types/priceIntelligence.ts
export type PriceSourceType =
  | "MANDI_DIRECT_TRANSACTION"
  | "GOVERNMENT_DATA"
  | "EXTERNAL_API"
  | "ADMIN_IMPORT"
  | "OTHER";

export type PriceTrend =
  | "RISING"
  | "FALLING"
  | "STABLE"
  | "VOLATILE"
  | "INSUFFICIENT_DATA";

export type ConfidenceLevel = "LOW" | "MEDIUM" | "HIGH";

export type ExpectedPriceComparisonState =
  | "BELOW"
  | "WITHIN"
  | "ABOVE"
  | "NOT_SPECIFIED";

export type MatchLevel =
  | "PRODUCT_VARIETY_QUALITY_LOCATION"
  | "PRODUCT_VARIETY_QUALITY"
  | "PRODUCT_QUALITY"
  | "PRODUCT_ONLY"
  | "CATEGORY_ONLY"
  | "INSUFFICIENT_DATA";

export interface PriceObservation {
  id: string;
  product_name: string;
  category: string;
  variety?: string | null;
  quality_grade?: string | null;
  price: number;
  currency: string;
  price_unit: string;
  market_name?: string | null;
  district?: string | null;
  state?: string | null;
  source_type: PriceSourceType;
  source_name: string;
  source_reference?: string | null;
  observation_date: string;
  price_per_kg: number;
  created_at: string;
  updated_at: string;
}

export interface PriceObservationListResponse {
  items: PriceObservation[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface PriceObservationCreateInput {
  product_name: string;
  category?: string;
  variety?: string;
  quality_grade?: string;
  price: number;
  currency?: string;
  price_unit?: string;
  market_name?: string;
  district?: string;
  state?: string;
  source_type: PriceSourceType;
  source_name: string;
  source_reference?: string;
  observation_date: string;
}

export interface PriceObservationUpdateInput {
  product_name?: string;
  category?: string;
  variety?: string;
  quality_grade?: string;
  price?: number;
  currency?: string;
  price_unit?: string;
  market_name?: string;
  district?: string;
  state?: string;
  source_type?: PriceSourceType;
  source_name?: string;
  source_reference?: string;
  observation_date?: string;
}

export interface CSVImportRowError {
  row_number: number;
  error: string;
  raw_data?: Record<string, any>;
}

export interface CSVImportResponse {
  total_rows: number;
  imported_count: number;
  error_count: number;
  errors: CSVImportRowError[];
}

export interface PriceIntelligenceSummary {
  product_name: string;
  category: string;
  variety?: string | null;
  quality_grade?: string | null;
  latest_price?: number | null;
  min_price?: number | null;
  max_price?: number | null;
  avg_price?: number | null;
  median_price?: number | null;
  weighted_avg_price?: number | null;
  avg_7d?: number | null;
  avg_30d?: number | null;
  avg_90d?: number | null;
  trend: PriceTrend;
  confidence: ConfidenceLevel;
  observation_count: number;
  sources: string[];
  disclaimer: string;
}

export interface PriceHistoryPoint {
  observation_date: string;
  avg_price_per_kg: number;
  min_price_per_kg: number;
  max_price_per_kg: number;
  observation_count: number;
  source_types: string[];
}

export interface PriceHistoryResponse {
  product_name: string;
  variety?: string | null;
  quality_grade?: string | null;
  start_date: string;
  end_date: string;
  history: PriceHistoryPoint[];
  total_observations: number;
  overall_avg_per_kg?: number | null;
  overall_min_per_kg?: number | null;
  overall_max_per_kg?: number | null;
  trend: PriceTrend;
}

export interface RegionalPriceItem {
  region_type: string;
  region_name: string;
  avg_price_per_kg?: number | null;
  min_price_per_kg?: number | null;
  max_price_per_kg?: number | null;
  observation_count: number;
  status_message: string;
}

export interface RegionalPriceResponse {
  product_name: string;
  variety?: string | null;
  quality_grade?: string | null;
  district?: string | null;
  state?: string | null;
  district_reference?: RegionalPriceItem | null;
  state_reference?: RegionalPriceItem | null;
  mandi_direct_reference?: RegionalPriceItem | null;
  overall_range_text: string;
  has_sufficient_data: boolean;
}

export interface PriceRecommendationRequest {
  product_name: string;
  category?: string;
  variety?: string;
  quality_grade?: string;
  district?: string;
  state?: string;
  expected_price?: number;
  price_unit?: string;
}

export interface PriceRecommendationResponse {
  product_name: string;
  variety?: string | null;
  quality_grade?: string | null;
  reference_price: number;
  recommended_min_price: number;
  recommended_target_price: number;
  recommended_max_price: number;
  display_unit: string;
  display_reference_price: number;
  display_min_price: number;
  display_target_price: number;
  display_max_price: number;
  trend: PriceTrend;
  confidence: ConfidenceLevel;
  match_level: MatchLevel;
  observation_count: number;
  data_sources: string[];
  explanation: string;
  expected_price_comparison: ExpectedPriceComparisonState;
  comparison_message: string;
  disclaimer: string;
  generated_at: string;
}

File: mandi-direct/backend/app/schemas/price_intelligence.py
from datetime import date, datetime
from decimal import Decimal
from enum import Enum
from typing import List, Optional
from pydantic import BaseModel, Field, field_validator


class PriceSourceType(str, Enum):
    MANDI_DIRECT_TRANSACTION = "MANDI_DIRECT_TRANSACTION"
    GOVERNMENT_DATA = "GOVERNMENT_DATA"
    EXTERNAL_API = "EXTERNAL_API"
    ADMIN_IMPORT = "ADMIN_IMPORT"
    OTHER = "OTHER"


class PriceTrend(str, Enum):
    RISING = "RISING"
    FALLING = "FALLING"
    STABLE = "STABLE"
    VOLATILE = "VOLATILE"
    INSUFFICIENT_DATA = "INSUFFICIENT_DATA"


class ConfidenceLevel(str, Enum):
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"


class ExpectedPriceComparisonState(str, Enum):
    BELOW = "BELOW"
    WITHIN = "WITHIN"
    ABOVE = "ABOVE"
    NOT_SPECIFIED = "NOT_SPECIFIED"


class MatchLevel(str, Enum):
    EXACT_FULL = "PRODUCT_VARIETY_QUALITY_LOCATION"
    VARIETY_QUALITY = "PRODUCT_VARIETY_QUALITY"
    QUALITY_ONLY = "PRODUCT_QUALITY"
    PRODUCT_ONLY = "PRODUCT_ONLY"
    CATEGORY_ONLY = "CATEGORY_ONLY"
    INSUFFICIENT_DATA = "INSUFFICIENT_DATA"


# --- Price Observation CRUD Schemas ---

class PriceObservationBase(BaseModel):
    product_name: str = Field(..., max_length=150, example="Tomato")
    category: str = Field(default="VEGETABLE", max_length=50)
    variety: Optional[str] = Field(default=None, max_length=100, example="Hybrid")
    quality_grade: Optional[str] = Field(default="UNGRADED", max_length=30, example="GRADE_A")
    price: Decimal = Field(..., gt=0, description="Observed price per specified price_unit")
    currency: str = Field(default="INR", max_length=10)
    price_unit: str = Field(default="PER_KG", max_length=20, example="PER_KG")
    market_name: Optional[str] = Field(default=None, max_length=150, example="Azadpur Mandi")
    district: Optional[str] = Field(default=None, max_length=100, example="Kolar")
    state: Optional[str] = Field(default=None, max_length=100, example="Karnataka")
    source_type: PriceSourceType = Field(..., description="Source of observation")
    source_name: str = Field(..., max_length=150, example="Agmarknet Portal")
    source_reference: Optional[str] = Field(default=None, max_length=255)
    observation_date: date = Field(..., example="2026-09-10")

    @field_validator("price")

    def validate_positive_price(cls, v: Decimal) -> Decimal:
        if v <= 0:
            raise ValueError("Price must be strictly greater than 0")
        return v.quantize(Decimal("0.01"))


class PriceObservationCreate(PriceObservationBase):
    pass


class PriceObservationUpdate(BaseModel):
    product_name: Optional[str] = Field(default=None, max_length=150)
    category: Optional[str] = Field(default=None, max_length=50)
    variety: Optional[str] = Field(default=None, max_length=100)
    quality_grade: Optional[str] = Field(default=None, max_length=30)
    price: Optional[Decimal] = Field(default=None, gt=0)
    currency: Optional[str] = Field(default=None, max_length=10)
    price_unit: Optional[str] = Field(default=None, max_length=20)
    market_name: Optional[str] = Field(default=None, max_length=150)
    district: Optional[str] = Field(default=None, max_length=100)
    state: Optional[str] = Field(default=None, max_length=100)
    source_type: Optional[PriceSourceType] = None
    source_name: Optional[str] = Field(default=None, max_length=150)
    source_reference: Optional[str] = Field(default=None, max_length=255)
    observation_date: Optional[date] = None


class PriceObservationResponse(PriceObservationBase):
    id: str
    price_per_kg: Decimal = Field(..., description="Normalized price in INR per KG")
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class PriceObservationListResponse(BaseModel):
    items: List[PriceObservationResponse]
    total: int
    page: int
    size: int
    pages: int


# --- CSV Import Schemas ---

class CSVImportRowError(BaseModel):
    row_number: int
    error: str
    raw_data: Optional[dict] = None


class CSVImportResponse(BaseModel):
    total_rows: int
    imported_count: int
    error_count: int
    errors: List[CSVImportRowError]


# --- Price Intelligence Summary & History Schemas ---

class PriceIntelligenceSummary(BaseModel):
    product_name: str
    category: str
    variety: Optional[str] = None
    quality_grade: Optional[str] = None
    latest_price: Optional[Decimal] = None
    min_price: Optional[Decimal] = None
    max_price: Optional[Decimal] = None
    avg_price: Optional[Decimal] = None
    median_price: Optional[Decimal] = None
    weighted_avg_price: Optional[Decimal] = None
    avg_7d: Optional[Decimal] = None
    avg_30d: Optional[Decimal] = None
    avg_90d: Optional[Decimal] = None
    trend: PriceTrend
    confidence: ConfidenceLevel
    observation_count: int
    sources: List[str]
    disclaimer: str = "Price Intelligence provides market guidance based on available data and is not a guaranteed selling price."


class PriceHistoryPoint(BaseModel):
    observation_date: date
    avg_price_per_kg: Decimal
    min_price_per_kg: Decimal
    max_price_per_kg: Decimal
    observation_count: int
    source_types: List[str]


class PriceHistoryResponse(BaseModel):
    product_name: str
    variety: Optional[str] = None
    quality_grade: Optional[str] = None
    start_date: date
    end_date: date
    history: List[PriceHistoryPoint]
    total_observations: int
    overall_avg_per_kg: Optional[Decimal] = None
    overall_min_per_kg: Optional[Decimal] = None
    overall_max_per_kg: Optional[Decimal] = None
    trend: PriceTrend


# --- Regional Price Schemas ---

class RegionalPriceItem(BaseModel):
    region_type: str = Field(..., description="DISTRICT | STATE | MANDI_DIRECT_TRANSACTION | ALL")
    region_name: str
    avg_price_per_kg: Optional[Decimal] = None
    min_price_per_kg: Optional[Decimal] = None
    max_price_per_kg: Optional[Decimal] = None
    observation_count: int
    status_message: str


class RegionalPriceResponse(BaseModel):
    product_name: str
    variety: Optional[str] = None
    quality_grade: Optional[str] = None
    district: Optional[str] = None
    state: Optional[str] = None
    district_reference: Optional[RegionalPriceItem] = None
    state_reference: Optional[RegionalPriceItem] = None
    mandi_direct_reference: Optional[RegionalPriceItem] = None
    overall_range_text: str
    has_sufficient_data: bool


# --- Recommendation Schemas ---

class PriceRecommendationRequest(BaseModel):
    product_name: str = Field(..., max_length=150, example="Tomato")
    category: Optional[str] = Field(default=None, max_length=50)
    variety: Optional[str] = Field(default=None, max_length=100)
    quality_grade: Optional[str] = Field(default=None, max_length=30)
    district: Optional[str] = Field(default=None, max_length=100)
    state: Optional[str] = Field(default=None, max_length=100)
    expected_price: Optional[Decimal] = Field(default=None, gt=0, description="Farmer's proposed price per price_unit")
    price_unit: str = Field(default="PER_KG", max_length=20, example="PER_KG")


class PriceRecommendationResponse(BaseModel):
    product_name: str
    variety: Optional[str] = None
    quality_grade: Optional[str] = None
    reference_price: Decimal = Field(..., description="Calculated market reference price per KG")
    recommended_min_price: Decimal = Field(..., description="Recommended fair range floor per KG")
    recommended_target_price: Decimal = Field(..., description="Suggested target price per KG")
    recommended_max_price: Decimal = Field(..., description="Recommended fair range ceiling per KG")
    
    # Original unit converted prices (for display in farmer's selected unit e.g. PER_QUINTAL)
    display_unit: str
    display_reference_price: Decimal
    display_min_price: Decimal
    display_target_price: Decimal
    display_max_price: Decimal

    trend: PriceTrend
    confidence: ConfidenceLevel
    match_level: MatchLevel
    observation_count: int
    data_sources: List[str]
    explanation: str
    
    expected_price_comparison: ExpectedPriceComparisonState
    comparison_message: str
    
    disclaimer: str = "Price Intelligence provides market guidance based on available data and is not a guaranteed selling price."
    generated_at: datetime
